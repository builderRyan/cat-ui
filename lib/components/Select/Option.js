"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var react_1 = require("react");
var classnames_1 = require("classnames");
var styled_1 = require("./styled");
var Option = (function (_super) {
    __extends(Option, _super);
    function Option() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Option.prototype.render = function () {
        var _a = this.props, className = _a.className, style = _a.style, children = _a.children;
        var classes = classnames_1.default('hmly-select', className);
        return (React.createElement(styled_1.StyledOption, { className: classes, style: style }, children));
    };
    return Option;
}(react_1.Component));
exports.default = Option;
//# sourceMappingURL=Option.js.map