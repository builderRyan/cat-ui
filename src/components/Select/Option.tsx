import * as React from 'react';
import { Component } from 'react';
import classNames from 'classnames';

export interface OptionProps {
  /** 类名 */
  className?: string;
  /** 样式 */
  style?: object;
  /** 当前条目的值 */
  value?: string | number;
  /** 是否禁用 */
  disable?: boolean;
  /** 内部点击事件的回调 */
  onOptionClick?: any;
  /** 外部点击事件的回调 */
  onClick?: (value?: string | number, child?: string | React.ReactNode) => {};
}

/**
 * Option
 */
class Option extends Component<OptionProps, any> {
  static defaultProps = {
    disable: false
  };

  handleClick = () => {
    const { value, onOptionClick, onClick, disable, children } = this.props;
    if (!disable && onOptionClick) {
      onOptionClick(value, children);
    }
    onClick && onClick(value, children);
  };

  render() {
    const { className, style, disable, children } = this.props;
    const classes = classNames(
      'cat-option',
      {
        'cat-option--disable': disable
      },
      className
    );

    return (
      <div className={classes} style={style} onClick={this.handleClick}>
        {children}
      </div>
    );
  }
}

export default Option;
