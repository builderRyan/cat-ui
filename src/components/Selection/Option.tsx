import * as React from 'react'
import * as PropTypes from 'prop-types';
import { PureComponent } from 'react'
import classNames from 'classnames'

interface IOptionInfo {
  /**
   * 当前项的值
   */
  value?: string | number;
   /**
   * 当前项的索引
   */
  index?: number;
   /**
   * 当前项的索引
   */
  disabled?: boolean;
}

export interface IOptionProps {
  /**
  * 样式前缀
  */
  prefix?: string;
  /**
   * css 类名
   */
  className?: string;
  /**
   * 行内样式
   */
  style?: object;
  /**
   * 当前项的值
   */
  value?: string | number;
  /**
   * 当前项的索引
   */
  index?: number;
  /**
   * 禁用该项
   */
  disabled?: boolean;
  /**
   * react element
   */
  children?: React.ReactNode;
  /**
   * 点击的回调
   */
  onClick?: (optionInfo: IOptionInfo) => {};
  /**
   * 鼠标移入的回调
   */
  onMouseEnter?: (optionInfo: IOptionInfo) => {};
  /**
   * 鼠标移出的回调
   */
  onMouseLeave?: (optionInfo: IOptionInfo) => {};
}

/**
 * Option组件： 一个可选项的容器组件
 * 服务于Selection
 */
class Option extends PureComponent<IOptionProps, {}> {
  static propTypes = {
    prefix: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.object,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
    index: PropTypes.number,
    disabled: PropTypes.bool,
    children: PropTypes.node,
    onClick: PropTypes.func,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
  };

  static defaultProps = {
    prefix: 'cat'
  };

  constructor(props: IOptionProps) {
    super(props)
  }

  handleClick = () => {
    const { value, index, disabled, onClick } = this.props
    onClick && onClick({value, index, disabled})
  }

  handleMouseEnter = () => {
    const { value, index, disabled, onMouseEnter } = this.props
    onMouseEnter && onMouseEnter({value, index, disabled})
  }

  handleMouseLeave = () => {
    const { value, index, disabled, onMouseLeave } = this.props
    onMouseLeave && onMouseLeave({value, index, disabled})
  }

  render () {
    const { prefix, style, className, value, children } = this.props
    const classes = classNames(`${prefix}-select-option`, className)

    return (
      <div
        className={classes}
        style={style}
        onClick={this.handleClick}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
      >
        {children || value}
      </div>
    )
  }
}

export default Option;