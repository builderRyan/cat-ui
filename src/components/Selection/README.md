# Selection 选择选项

## Example

```jsx
<Selection
  open={openSelection}
  value='选项 2'
  onSelect={this.handleSelect}
>
  <Option className='qwer' value='选项 1' />
  <Option value='选项 2' />
  <Option value='选项 3' />
  <Option value='选项 4' />
  <Option value='选项 5' />
</Selection>
```

## Selection API

|    属性     |         类型          |    默认值    |                描述                | 是否必填 |
| :---------: | :-------------------: | :----------: | :--------------------------------: | :------: |
|   prefix    |       `string`        | `cat-avatar` |              样式前缀              |    -     |
|  className  |       `string`        |      -       |                类名                |    -     |
|    style    |       `object`        |      -       |              行内样式              |    -     |
|    open    |       `boolean`        |     true     |         是否显示选择菜单            |    -     |
|  defaultValue  |       `string`     |      -       |         用于初始化的默认值          |    -     |
|    value    |       `string`        |      -       |              选中项的值            |    -     |
|   children  |     `ReactElement`    |      -       |              React元素             |    -     |
|   onSelect  |       `function`      |      -       |        选中一项触发的回调函数        |    -     |

## Option API

|    属性     |         类型          |    默认值    |                描述                | 是否必填 |
| :---------: | :-------------------: | :----------: | :--------------------------------: | :------: |
|   prefix    |       `string`        | `cat-avatar` |              样式前缀               |    -     |
|  className  |       `string`        |      -       |                类名                 |    -     |
|    style    |       `object`        |      -       |              行内样式               |    -     |
|    value    |       `string`        |      -       |              选中项的值             |    -     |
|    index    |       `number`        |      -       |              选中项的索引           |    -     |
|  disabled   |       `bool`          |      -       |              禁用该项（不可选）      |    -     |
|  children   | `string或ReactElement` |      -       |              react元素             |    -     |
|   onClick   |       `function`      |      -       |              点击事件               |    -     |
| onMouseEnter |       `function`     |      -       |             鼠标移入事件            |    -     |
| onMouseLeave |       `function`     |      -       |             鼠标移出事件            |    -     |
