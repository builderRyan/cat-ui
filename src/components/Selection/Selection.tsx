import * as React from 'react'
import * as PropTypes from 'prop-types'
import keycode from 'keycode'
import { PureComponent, cloneElement } from 'react'
import classNames from 'classnames'

export interface ISelectionProps {
  /**
   * 样式前缀
   */
  prefix?: string;
  /**
   * css 类名
   */
  className?: string;
  /**
   * 行内样式
   */
  style?: object;
  /**
   * 是否显示选择菜单
   */
  visible?: boolean;
  /**
   * 用于初始化的默认值
   */
  defaultValue?: string | number;
  /**
   * 指定当前选中项
   */
  value?: string | number;
  /**
   * 子组件
   */
  children?: any;
  /**
   * 选中一项触发的回调函数
   */
  onSelect?: (value?: string | number) => void;
}

export interface ISelectionState {
  /**
   * 是否显示当前组件
   */
  visible: boolean,
  /**
   * 获得焦点项的索引
   */
  focusIndex?: number,
  /**
   * 选中项的索引
   */
  selectedIndex?: number
}

/**
 * 选择菜单组件
 * 应用在具有选择的场景
 * 比如Select、Search、列表等
 */
class Selection extends PureComponent<ISelectionProps, ISelectionState> {
  static propTypes = {
    prefix: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.object,
    visible: PropTypes.bool,
    defaultValue: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
    children: PropTypes.node,
    onSelect: PropTypes.func
  };

  static defaultProps = {
    visible: true,
    prefix: 'cat'
  }

  constructor(props: ISelectionProps) {
    super(props)
    const { visible, defaultValue, value, children } = props
    const initialValue = defaultValue || value
    let focusIndex
    
    React.Children.forEach(children, (Option, index) => {
      if (Option.props.value === initialValue) {
        return focusIndex = index
      }
    })

    this.state = {
      visible,
      focusIndex,
      selectedIndex: focusIndex,
    }
  }

  private childrenLength = 0

  componentDidMount () {
    document.addEventListener('keydown', this.handleKeydown)
  }

  componentWillUnmount () {
    document.removeEventListener('keydown', this.handleKeydown)
  }

  static getDerivedStateFromProps(nextProps: ISelectionProps, prevState: ISelectionState) {
    if (nextProps.visible !== prevState.visible) {
      return {
        visible: nextProps.visible,
        focusIndex: -1
      }
    }
    return null
  }

  getValidItemIndex = (itemIndex: number, offset: number) => {
    const { children } = this.props

    let nextIndex = itemIndex
    while (children[nextIndex].props.disabled) {
      nextIndex += offset
    }

    return nextIndex
  }

  moveFocusIndex = (offset: number) => {
    const { focusIndex } = this.state
    let nextIndex = focusIndex + offset
  
    if (nextIndex === this.childrenLength) {
      nextIndex = 0
    } else if (nextIndex < 0) {
      nextIndex = this.childrenLength - 1
    }

    this.setState({
      focusIndex: this.getValidItemIndex(nextIndex, offset)
    })
  }

  handleClickOption = ({value, index, disabled}) => {
    const { onSelect } = this.props

    if (disabled) {
      return 
    }

    this.setState({
      selectedIndex: index,
      focusIndex: -1
    })
    onSelect && onSelect(value)
  }

  handleMouseEnterOption = ({index, disabled}) => {
    if (disabled) {
      return
    }

    this.setState({
      focusIndex: index
    })
  }

  handleMouseLeaveOption = ({index, disabled}) => {
    if (disabled) {
      return
    }

    this.setState({
      focusIndex: index
    })
  }

  selectCurrentFocusIndex = () => {
    const { onSelect, children } = this.props
    const { focusIndex } = this.state

    if (!children[focusIndex]) {
      return
    }

    let value = children[focusIndex].props.value
    let disabled = children[focusIndex].props.disabled

    if (disabled) {
      return
    }

    this.setState({
      selectedIndex: focusIndex
    })
    onSelect && onSelect(value)
  }

  handleKeydown = e => {
    const { visible } = this.state
    const keyName = keycode(e)

    if (!visible) {
      return
    }

    switch (keyName) {
      case 'down':
        e.preventDefault()
        this.moveFocusIndex(1)
        break;

      case 'up': {
        e.preventDefault();
        this.moveFocusIndex(-1)
        break;
      }

      case 'enter': {
        this.selectCurrentFocusIndex()
        break;
      }

      default:
    }
  }

  render () {
    const { prefix, style, className, children } = this.props
    const { visible, focusIndex, selectedIndex } = this.state
    const classes = classNames(`${prefix}-selection-wrap`, {
      'visible': visible,
    }, className);

    this.childrenLength = 0
    const Options = React.Children.map(children, (Option: React.ReactElement, index) => {
      if (!('value' in Option.props)) {
        return null
      }

      this.childrenLength++

      return cloneElement(Option, {
        className: classNames(Option.props.className, {
          'disabled': Option.props.disabled,
          'focus': index === focusIndex,
          'selected': index === selectedIndex
        }),
        index,
        disabled: Option.props.disabled,
        onClick: this.handleClickOption,
        onMouseEnter: this.handleMouseEnterOption,
        onMouseLeave: this.handleMouseLeaveOption,
      })
    })

    return (
      <div
        className={classes}
        style={style}
        onKeyDown={this.handleKeydown}
      >
        {Options}
      </div>
    )
  }
}

export default Selection;