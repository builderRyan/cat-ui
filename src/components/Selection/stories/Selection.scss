.cat-selection-wrap {
  width: 189px;
}