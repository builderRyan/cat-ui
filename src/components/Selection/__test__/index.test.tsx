import * as React from 'react';
import * as Enzyme from 'enzyme';
import { shallow, mount } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import Selection from '../index';
import Option from '../Option';
import ReactTestUtils from 'react-dom/test-utils'; 

describe('Option', () => {
  beforeEach(() => {
    document.querySelector('body').innerHTML = '';
  });
  test('option disabled', () => {
    const wrapper = mount(
      <Selection
        visible={false}
        defaultValue='选项 1'
        className='as'
        onSelect={this.handleSelect}
      >
        <Option disabled value='选项 1' />
        <Option value='选项 2' />
        <Option value='选项 3' />
      </Selection>
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Option', () => {
  beforeEach(() => {
    document.querySelector('body').innerHTML = '';
  });
  test('option mouse event', () => {
    const wrapper = mount(
      <Selection
        visible
        defaultValue='选项 1'
        className='as'
        onSelect={this.handleSelect}
      >
        <Option className='my-option' value='选项 1'>
          <span id='span'>123</span>
        </Option>
        <Option value='选项 2' />
        <Option value='选项 3' />
      </Selection>
    );

    wrapper.find('.my-option').at(0).simulate('mouseenter');
    wrapper.find('.my-option').at(0).simulate('mouseleave');
    wrapper.find('.my-option').at(0).simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Selection', () => {
  test('renders correctly', () => {
    const wrapper = mount(
      <Selection
        visible
        value='选项 2'
        onSelect={this.handleSelect}
      >
        <Option className='option1' value='选项 1' />
        <Option value='选项 2' />
        <Option value='选项 3' />
        <Option value='选项 4' />
        <Option value='选项 5' />
        <Option />
        <Option>
          <span id='custom'>custom option</span>
        </Option>
      </Selection>
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Selection', () => {
  beforeEach(() => {
    document.querySelector('body').innerHTML = '';
  });
  test('keyboard event', () => {
    const wrapper = mount(
      <Selection
        visible
        value='选项 2'
        defaultValue='选项 1'
        className='my-selection'
        onSelect={this.handleSelect}
      >
        <Option className='option-1' value='选项 1' />
        <Option disabled className='option-2' value='选项 2' />
        <Option className='option-3' value='选项 3' />
      </Selection>
    );
    wrapper.find('.option-1').at(0).simulate('mouseenter');
    wrapper.find('.option-1').at(0).simulate('mouseleave');
    wrapper.find('.option-2').at(0).simulate('mouseenter');


    wrapper.find('.option-1').at(0).simulate('mouseenter');
    const event1 = new KeyboardEvent('keydown', {key: 'Enter'});
    document.dispatchEvent(event1);

    wrapper.find('.option-2').at(0).simulate('mouseenter');
    wrapper.find('.option-2').at(0).simulate('click');
    wrapper.find('.option-2').at(0).simulate('mouseleave');

    wrapper.find('.option-3').at(0).simulate('mouseenter');
    wrapper.find('.my-selection').at(0).simulate('keydown', {key: 'Down'});
    // const event2 = new KeyboardEvent('keydown', {key: 'Down'});
    // document.dispatchEvent(event2);


    wrapper.find('.option-2').at(0).simulate('mouseenter');
    const event3 = new KeyboardEvent('keydown', {key: 'Enter'});
    document.dispatchEvent(event3);

    expect(wrapper).toMatchSnapshot();
  });
});