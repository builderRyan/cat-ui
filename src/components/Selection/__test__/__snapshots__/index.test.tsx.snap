// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selection renders correctly 1`] = `
<Selection
  open={true}
  prefix="cat"
  value="选项 2"
>
  <div
    className="cat-selection-wrap"
    onKeyDown={[Function]}
  >
    <Option
      className="qwer"
      index={0}
      key="0/.0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      prefix="cat"
      value="选项 1"
    >
      <div
        className="cat-select-option qwer"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        选项 1
      </div>
    </Option>
    <Option
      className="focus selected"
      index={1}
      key="1/.1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      prefix="cat"
      value="选项 2"
    >
      <div
        className="cat-select-option focus selected"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        选项 2
      </div>
    </Option>
    <Option
      className=""
      index={2}
      key="2/.2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      prefix="cat"
      value="选项 3"
    >
      <div
        className="cat-select-option"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        选项 3
      </div>
    </Option>
    <Option
      className=""
      index={3}
      key="3/.3"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      prefix="cat"
      value="选项 4"
    >
      <div
        className="cat-select-option"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        选项 4
      </div>
    </Option>
    <Option
      className=""
      index={4}
      key="4/.4"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      prefix="cat"
      value="选项 5"
    >
      <div
        className="cat-select-option"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        选项 5
      </div>
    </Option>
  </div>
</Selection>
`;
